File: app/src/at-phuongle/java/com/asiantech/intern20summer1/intro/EdgeOfTheOcean.java
package com.asiantech.intern20summer1.intro;

public class EdgeOfTheOcean {
    /**
     * Write a function that sort array from min to max
     */
    void sortASC(int[] arr) {
        int temp = arr[0];
        for (int i = 0; i < arr.length - 1; i++) {
            for (int j = i + 1; j < arr.length; j++) {
                if (arr[i] > arr[j]) {
                    temp = arr[j];
                    arr[j] = arr[i];
                    arr[i] = temp;
                }
            }
        }
    }

    /**
     * Write a function that check the palindrome string
     */
    int adjacentElementsProduct(int[] inputArray) {
        int size = inputArray.length;
        int n = inputArray[0] * inputArray[1];
        for (int i = 1; i < size - 1; i++) {
            int j = i + 1;
            if ((inputArray[i] * inputArray[j]) > n) {
                n = inputArray[i] * inputArray[j];
            }
        }
        return n;
    }

    /**
     * Write a function that returns the area of the polygons have edges n
     */
    int shapeArea(int n) {
        return n * n + (n - 1) * (n - 1);
    }

    /**
     * Write a function that returns number of elements to become a continuous array
     */
    int makeArrayConsecutive2(int[] statues) {
        int n = 0;
        sortASC(statues);
        for (int i = 0; i < statues.length - 1; i++) {
            n += (statues[i + 1] - statues[i] - 1);
        }
        return n;
    }

    /**
     * Write a function that checks the given sequence is an increasing sequence when remove 1 element
     */
    boolean almostIncreasingSequence(int[] sequence) {
        int n = 0;

        for (int i = 0; i < sequence.length - 1; i++) {
            if (sequence[i] - sequence[i + 1] >= 0) {
                n++;
                if (i - 1 >= 0 && i + 2 <= sequence.length - 1
                        && sequence[i] - sequence[i + 2] >= 0
                        && sequence[i - 1] - sequence[i + 1] >= 0) {
                    return false;
                }
            }
        }
        return n <= 1;
    }

    /**
     * Write a function that returns the total sum of all rooms that having the values don't appear below 0
     */
    int matrixElementsSum(int[][] matrix) {
        int sum = 0;
        for (int i = 0; i < matrix[0].length; i++) {
            for (int j = 0; j < matrix.length; j++) {
                if (matrix[j][i] > 0)
                    sum += matrix[j][i];
                else
                    break;
            }
        }
        return sum;
    }
}

File: app/src/at-phuongle/java/com/asiantech/intern20summer1/intro-kotlin/EdgeOfTheOcean.kt
package com.asiantech.intern20summer1.`intro-kotlin`

object EdgeOfTheOcean {
    fun adjacentElementsProduct(inputArray : Array<Int>) : Int {
        var size : Int = inputArray.size;
        var n : Int = inputArray.get(0) * inputArray.get(1);
        for (i in 1 until (size-1)) {
            var j : Int = i + 1;
            var temp : Int = inputArray.get(i) * inputArray.get(j)
            if (temp > n) {
                n = inputArray.get(i) * inputArray.get(j);
            }
        }
        return n;
    }

    fun shapeArea(n : Int) : Int {
        return n * n + (n - 1) * (n - 1);
    }

    fun sortASC(arr : Array<Int>) : Unit {
        var temp : Int = arr[0];
        for (i in 0 until (arr.size - 1)) {
            for (j in (i+1) until (arr.size)) {
                if (arr[i] > arr[j]) {
                    temp = arr[j];
                    arr[j] = arr[i];
                    arr[i] = temp;
                }
            }
        }
    }

    fun makeArrayConsecutive2(statues : Array<Int>) : Int {
        var n : Int = 0;

        sortASC(statues);

        for (i in 0 until (statues.size-1)) {
            n += (statues[i + 1] - statues[i] - 1);
        }

        return n;
    }

    fun almostIncreasingSequence(sequence : Array<Int>) : Boolean {
        var n : Int = 0;

        for(i in 0 until (sequence.size - 1)){

            if(sequence[i] - sequence[i+1] >= 0){
                n++;
                if(i - 1 >= 0 && i + 2 <= sequence.size - 1
                    && sequence[i] - sequence[i+2] >= 0
                    && sequence[i-1] - sequence[i+1] >= 0){
                    return false;
                }
            }
        }

        return n <= 1;
    }

}